### PR TITLE
[#71] [Android] [Integrate] As a user, I can quit the survey

### LINKED ISSUE #71: [Android] [Integrate] As a user, I can quit the survey
## Why

When a user decides to cancel/dismiss an ongoing survey, he/she must be asked to confirm before dismissing it. This is to prevent any unintentional dismissal of ongoing survey. Otherwise, if a user loses the progress of his/her survey, there are high chances that it might upset the user and unhappy users will unlikely re-take the survey.

It is important to note that if there is no progress on the survey, the mobile app should dismiss the screen without any prompt since the user might just want to see a preview of the questions or start the survey by mistake.

## Acceptance Criteria
Upon tapping the Dismiss button,
- If a user has replied to at least 1 question of the survey, prompt a modal dialog with the following content:
    - Title - Warning
    - Message - Are you sure you want to quit the survey?
    - Yes action - Yes
        - Selecting Yes dismisses the screen
    - Cancel action - Cancel
        - Selecting Cancel does nothing
- If the user HAS NOT replied to any survey question, dismiss the screen immediately.

## Design

Image

[Take Survey](https://www.figma.com/proto/GjRPOjDyZ6f4EDL3wKarRK/Challenge-Mobile-App?page-id=0%3A1&node-id=34%3A765&scaling=min-zoom)

### PR DESCRIPTION
**Pull request by suho**
- Resolves #71

## What happened 👀

Added exit alert dialog

## Insight 📝

I will write both unit and UI tests for this dialog after this [pull request](https://github.com/suho/kmm-ic/pull/123)

## Proof Of Work 📹

Screenshot 2022-12-15 at 15 44 43

### CODE REVIEW

**Review by hoangnguyen92dn — COMMENTED** [review]
I am curious if we could have UnitTest for the logic show/hide exit dialog? 😄

**Comment by suho** [comment]
> I am curious if we could have UnitTest for the logic show/hide exit dialog? 😄

@hoangnguyen92dn Please ignore this. I decided to write both unit and UI tests for this dialog after this [pull request](https://github.com/suho/kmm-ic/pull/123)

File: android/src/main/res/values/strings.xml [hunk 1]
```diff
@@ -12,4 +12,8 @@
 
     <string name="start_survey">Start Survey</string>
     <string name="submit_button">Submit</string>
+    <string name="exit_warning_title">Warning!</string>
+    <string name="exit_detail_message">Are you sure you want to quit the survey</string>
```
Comment (hoangnguyen92dn) [inline android/src/main/res/values/strings.xml:16]: @suho Missing the `?` :harold:

image

**Review by hoangnguyen92dn — APPROVED** [review]
The rest LGTM 🚀

File: android/src/main/res/values/strings.xml (hunk 1 above)
Comment (suho) [reply android/src/main/res/values/strings.xml]: Fixed in https://github.com/suho/kmm-ic/pull/122/commits/c1ecbbaf030b5f34a2041a03badf77dec7a09315

File: android/src/main/java/co/nimblehq/ic/kmm/suv/android/ui/screens/surveyquestions/SurveyQuestionsViewModel.kt [hunk 2]
```diff
@@ -58,4 +61,16 @@ class SurveyQuestionsViewModel(
...
+        _showExitDialog.value = true
+    }
+
+    fun onExitDialogConfirm() {
+        _showExitDialog.value = false
+    }
+
+    fun onExitDialogDismiss() {
+        _showExitDialog.value = false
+    }
```
Comment (luongvo) [inline android/src/main/java/co/nimblehq/ic/kmm/suv/android/ui/screens/surveyquestions/SurveyQuestionsViewModel.kt:75]: Should we manage `showExitDialog` state in the Composable screen instead?
If we would like to keep the dialog action handling in VM, please add unit test for them 🙏

File: android/src/main/java/co/nimblehq/ic/kmm/suv/android/ui/screens/surveyquestions/SurveyQuestionsScreen.kt [hunk 3]
```diff
@@ -42,6 +40,7 @@ fun SurveyQuestionsScreen(
     val errorMessage by viewModel.errorMessage.collectAsState()
     val coverImageUrl by viewModel.coverImageUrl.collectAsState()
     val questionUiModels by viewModel.questionContentUiModels.collectAsState()
+    val showExitDialog by viewModel.showExitDialog.collectAsState()
```
Comment (luongvo) [inline android/src/main/java/co/nimblehq/ic/kmm/suv/android/ui/screens/surveyquestions/SurveyQuestionsScreen.kt:43]: Yup, we can manage its state here instead in VM

File: android/src/main/java/co/nimblehq/ic/kmm/suv/android/ui/screens/surveyquestions/SurveyQuestionsScreen.kt [hunk 4]
```diff
@@ -54,14 +53,26 @@ fun SurveyQuestionsScreen(
         )
     }
 
+    showExitDialog.let {
```
Comment (luongvo) [inline android/src/main/java/co/nimblehq/ic/kmm/suv/android/ui/screens/surveyquestions/SurveyQuestionsScreen.kt:56]: `showExitDialog` is a non-null state, right?

File: android/src/main/java/co/nimblehq/ic/kmm/suv/android/ui/screens/surveyquestions/SurveyQuestionsScreen.kt [hunk 5]
```diff
@@ -54,14 +53,26 @@ fun SurveyQuestionsScreen(
...
+        if (showExitDialog) {
+            ExitAlertDialog(
+                onConfirmClick = {
+                    viewModel.onExitDialogConfirm()
+                    onCloseClick()
```
Comment (luongvo) [inline android/src/main/java/co/nimblehq/ic/kmm/suv/android/ui/screens/surveyquestions/SurveyQuestionsScreen.kt:61]: `onCloseClick` event here is duplicated with the one in `SurveyQuestionsScreenContent` but their meanings are different: one goes through the confirmation dialog and another one points to the navigation action. I think we should have them under different names.

**Review by luongvo — COMMENTED** [review]
rest lgtm

File: android/src/main/java/co/nimblehq/ic/kmm/suv/android/ui/screens/surveyquestions/SurveyQuestionsViewModel.kt (hunk 2 above)
Comment (suho) [reply android/src/main/java/co/nimblehq/ic/kmm/suv/android/ui/screens/surveyquestions/SurveyQuestionsViewModel.kt]: Yep, we can manage that state in the composable screen. Fixed in https://github.com/suho/kmm-ic/pull/122/commits/34df858b7cdc74f4a9252b232270a14b2ef99b47

File: android/src/main/java/co/nimblehq/ic/kmm/suv/android/ui/screens/surveyquestions/SurveyQuestionsScreen.kt (hunk 3 above)
Comment (suho) [reply android/src/main/java/co/nimblehq/ic/kmm/suv/android/ui/screens/surveyquestions/SurveyQuestionsScreen.kt]: Fixed in https://github.com/suho/kmm-ic/pull/122/commits/34df858b7cdc74f4a9252b232270a14b2ef99b47

File: android/src/main/java/co/nimblehq/ic/kmm/suv/android/ui/screens/surveyquestions/SurveyQuestionsScreen.kt (hunk 4 above)
Comment (suho) [reply android/src/main/java/co/nimblehq/ic/kmm/suv/android/ui/screens/surveyquestions/SurveyQuestionsScreen.kt]: Fixed in https://github.com/suho/kmm-ic/pull/122/commits/34df858b7cdc74f4a9252b232270a14b2ef99b47

File: android/src/main/java/co/nimblehq/ic/kmm/suv/android/ui/screens/surveyquestions/SurveyQuestionsScreen.kt (hunk 5 above)
Comment (suho) [reply android/src/main/java/co/nimblehq/ic/kmm/suv/android/ui/screens/surveyquestions/SurveyQuestionsScreen.kt]: Fixed in https://github.com/suho/kmm-ic/pull/122/commits/34df858b7cdc74f4a9252b232270a14b2ef99b47

**Review by luongvo — APPROVED** [review]

**Pull request merged**